src/main/scala/mesosphere/marathon/api/akkahttp/v2/AppsController.scala

```diff
@@ -322,7 +331,36 @@ class AppsController(
     }
   }
 
-  private def listRunningTasks(appId: PathId)(implicit identity: Identity): Route = ???
+  private def listRunningTasks(appId: PathId)(implicit identity: Identity): Route = {
+    val maybeApp = groupManager.app(appId)
+    maybeApp.map { app =>
+      authorized(ViewRunSpec, app).apply {
+        val tasksF = instanceTracker.instancesBySpec flatMap { instancesBySpec =>
+          runningTasks(Set(appId), instancesBySpec)
+        }
+        onSuccess(tasksF) { tasks =>
+          complete(Json.obj("tasks" -> tasks.toRaml))
+        }
+      }
+    } getOrElse {
+      reject(Rejections.EntityNotFound.noApp(appId))
+    }
+  }
+
+  private def runningTasks(appIds: Set[PathId], instancesBySpec: InstancesBySpec): Future[Set[EnrichedTask]] = {
+    Source(appIds)
+      .filter(instancesBySpec.hasSpecInstances)
+      .mapAsync(1)(id => healthCheckManager.statuses(id).map(_ -> id))
+      .mapConcat {
+        case (health, id) =>
+          instancesBySpec.specInstances(id).flatMap { instance =>
```

Comment:
We already have the instances, don't we?

Reply:
Ok, so after the quick discussion I think it should be more like this.

First we should use async/await instead of `flatMap`:

```scala
async {
  val instances = await(instancesBySpec.specInstances(id))
  instance.tasksMap.values.map { task =>
    EnrichedTask(id, task, instance.agentInfo, health.getOrElse(instance.instanceId, Nil))
  }
}
```
I'm surprised that `mapConcat` has no troubles with the future here.

Then it seems we would not need the filtering since `instancesBySpec.specInstances` is already doing that. So, I'd first fetch the instances and then attach the health for each:

```scala
Future.sequence(appIds.map { appId =>
  async {
    val instances = await(instancesBySpec.specInstances(appId))
    val healthStatuses = await(healthCheckManager.statuses(appId))
    instances.flatMap { instance =>
      val healths: Seq[???] = healthStatuses.getOrElse(instance.id, Nil)
      instance.tasksMap.values.map { task => EnrichedTask(appId, task, instance.agentInfo, healths)
    }
  }
})
```

that seems to be easier to comprehend. I guess a stream could parallelize the flat mapping over the instances.

Reply:
Let's just introduce some state:

```scala
Future.sequence(appIds.map { appId =>
  async {
    val instances = await(instancesBySpec.specInstances(appId))
    val healthStatuses = await(healthCheckManager.statuses(appId))
    val tasksSetBuilder = Set.newBuilder[EnrichedTask]
    instances.foreach { instance =>
        val healths: Seq[???] = healthStatuses.getOrElse(instance.id, Nil)
        instance.tasksMap.valuesIterator { task =>
           tasksSetBuilder +=  EnrichedTask(appId, task, instance.agentInfo, healths)
        }
    }
    tasksSetBuilder.build()
  }
})
```